networkx/convert_matrix.py

```diff
@@ -1360,3 +1374,174 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
     return from_numpy_matrix(
         A, parallel_edges=parallel_edges, create_using=create_using
     )
+
+
+def generate_node_dataframe(
+    G: nx.Graph,
+    funcs: List[Callable],
+):
+    """
+    Return a pandas DataFrame representation of nodes and their metadata.
+
+    This function accepts a graph and a collection of user-defined functions
+    and returns a pandas DataFrame in which the nodes' are the DataFrame index,
+    the node metadata keys are column names,
+    and node metadata values are the values in the DataFrame.
+
+    ``funcs`` is an iterable of callables whose signature is
+
+        f(n, d) -> pd.Series
```

Comment:
Why should this be returning a ```pd.Series```? Isn't it just computing a single value? I don't know a huge amount about ```pd.Series```.  Is it better to create a bunch of Series and then combine them? Or is it better to collect the values into a single Series?  

Another way to ask this question:  Would it be better to move the ```pd.Series``` part of these functions outside the function? Then the user's function could focus on creating the value rather than packaging it.
Seems like the calling function knows the node name and can construct the Series for the user.

Reply:
It's a good point, @dschult. I think the reason why I've structured the requirement this way is to nudge end-users to using the community-wide standard and idiomatic PyData data structures. Something I've learned over the past few years is that targeting the "right" data structures can greatly simplify the kind of code we write, and I think tooling that explicitly nudges users towards the "right" data structures for a problem can help encourage better code writing altogether.

A series is nice for other reasons too. For example, it behaves like an upgraded dictionary; a collection of Series becomes a DataFrame. Underneath the hood, users can define functions that compute a whole bunch of values at one shot, such as in the `plot_graph_tensors.py` file where `node_metadata(n, d)` automagically converts all node metadata into a pandas Series, thereby giving a whole lot of values at one shot, or they can compute one value only (like in `ten_times_number(n, d)`).

Reply:
Interesting... I think a light clicked from the ```node_metadata(n, d)``` example where the Series contains many columns of data for one node. In my lingo (that os probably not quite correct) the pd.Series we create in this function is a "row Series" for that node... not a "column Series" (many nodes with one computed value).  I have typically thought of DataFrames as collections of "column Series", but it is very nice to realize that they are also collections of "row Series" as well.    Thanks!

Maybe you could change the wording below (in line 1388) to make explicit the idea that the pandas Series created in a function can consist of multiple values. The collection ```funcs``` stacks these Series to make a record for each node. [change the words to make them match the usual idioms for pandas/xarray.]  I suppose this might be made clear in the example by including ```"ycoord": d["y_coord"]``` in the dict as well... but I'm not sure what is the best presentation.

Reply:
I think this can be marked as resolved.